flang/test/Semantics/call09.f90

```diff
@@ -87,8 +87,7 @@ subroutine test1 ! 15.5.2.9(5)
     call s02(p) ! ok
     !ERROR: Actual procedure argument has interface incompatible with dummy argument 'p=': function results have distinct types: REAL(4) vs INTEGER(4)
     call s02(ip)
-    !ERROR: Actual argument associated with procedure pointer dummy argument 'p=' must be a pointer unless INTENT(IN)
-    call s02(procptr())
+    call s02(procptr()) ! ok
```

Comment:
The error is appropriate in this case.  See F'2023 subclause 15.5.2.10 paragraph 5: "If a dummy argument is a procedure pointer, the corresponding actual argument shall be a procedure pointer, a reference to a function that returns a procedure pointer, a reference to the intrinsic function NULL, or a valid target for the dummy pointer in a pointer assignment statement. If the actual argument is not a pointer, the dummy argument shall have INTENT (IN); if the actual argument is not a dummy argument it becomes pointer associated with the actual argument, otherwise it becomes pointer associated with the ultimate argument of the actual argument."

The actual argument is not a procedure pointer, it is a reference to a function that returns a procedure pointer.  So the corresponding dummy argument must have `INTENT(IN)`.

Reply:
Yeah, I read this paragraph too and was debating about it, so I see your point, but it just seems to me that no harm can be done. We have a few function test cases.

Reply:
There are NAG tests for this that would fail.

I don't see how an extension is justified here.  If an application cannot use `INTENT(IN)`, then it's expecting to be passed a modifiable procedure pointer, and a function result should indeed be an error.  If an application can use `INTENT(IN)`, then it should do so and avoid the issue.